shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,391 @@
+import org.json.JSONObject;
+
+import java.util.Arrays;
+import java.util.HashMap;
+
+import io.flutter.embedding.engine.dart.DartExecutor;
+import io.flutter.plugin.common.JSONMethodCodec;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+
+/**
+ * {@link TextInputChannel} is a platform channel between Android and Flutter that is used to
+ * communicate information about the user's text input.
+ * <p>
+ * When the user presses an action button like "done" or "next", that action is sent from Android
+ * to Flutter through this {@link TextInputChannel}.
+ * <p>
+ * When an input system in the Flutter app wants to show the keyboard, or hide it, or configure
+ * editing state, etc. a message is sent from Flutter to Android through this {@link TextInputChannel}.
+ * <p>
+ * {@link TextInputChannel} comes with a default {@link io.flutter.plugin.common.MethodChannel.MethodCallHandler}
+ * that parses incoming messages from Flutter. Register a {@link TextInputMethodHandler} to respond
+ * to standard Flutter text input messages.
+ */
+public class TextInputChannel {
+  @NonNull
+  public final MethodChannel channel;
+  @Nullable
+  private TextInputMethodHandler textInputMethodHandler;
+
+  private final MethodChannel.MethodCallHandler parsingMethodHandler = new MethodChannel.MethodCallHandler() {
+    @Override
+    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
+      if (textInputMethodHandler == null) {
+        // If no explicit TextInputMethodHandler has been registered then we don't
+        // need to forward this call to an API. Return.
+        return;
+      }
+
+      try {
+        String method = call.method;
+        Object args = call.arguments;
+        switch (method) {
+          case "TextInput.show":
+            textInputMethodHandler.show();
+            result.success(null);
+            break;
+          case "TextInput.hide":
+            textInputMethodHandler.hide();
+            result.success(null);
+            break;
+          case "TextInput.setClient":
+            final JSONArray argumentList = (JSONArray) args;
+            final int textInputClientId = argumentList.getInt(0);
+            final JSONObject jsonConfiguration = argumentList.getJSONObject(1);
+            textInputMethodHandler.setClient(textInputClientId, Configuration.fromJson(jsonConfiguration));
+            result.success(null);
+            break;
+          case "TextInput.setEditingState":
+            final JSONObject editingState = (JSONObject) args;
+            textInputMethodHandler.setEditingState(TextEditState.fromJson(editingState));
+            result.success(null);
+            break;
+          case "TextInput.clearClient":
+            textInputMethodHandler.clearClient();
+            result.success(null);
+            break;
+          default:
+            result.notImplemented();
+            break;
+        }
+      } catch (Exception exception) {
```

Comment:
Wrapping the entire case statement with all of the different method calls in a catch block and then catching `Exception` itself seems like a bad idea to me. We have no way of knowing if this was actually JSON problems or something more serious.

Reply:
I altered the catch to become a cascaded catch taht first catches a JSONException and reports it with the given message, then there is a catch-all for Exception where the error result simply outputs the exception message and therefore does not claim it's a JSON issue.

In either event, it seems that we don't ever want to crash the app in this location, so regardless of what goes wrong I think the thing we want to do is respond with an error back to Dart.

Reply:
On 2nd look, I added more exceptions for bad enum values so I've now refactored the exception catching to be per case statement.